share/vss/vss.go

```diff
@@ -0,0 +1,579 @@
+// vss.go implements the verifiable secret sharing scheme by Pedersen recalled
+// in "Provably Secure Distributed Schnorr Signatures and a (t, n) Threshold
+// Scheme for Implicit Certificates"  by Stinson and Strobl.
+// The basic scheme is taken from Pedersen's paper: "Non-interactive and
+// information-theoretic secure verifiable secret sharing", in Crypto ’91, pages
+// 129–140, 1991.
+// At a higher level, you have one party called the Dealer who wishes to share a
+// secret securly between other parties, called the Verifiers. In this scheme,
+// neither the verifiers or the dealer is deemed trustworthy, but rather, the
+// assumptions is that the adversary has at most t verifiers amongst n, or the
+// dealer is malicious. In that case, there needs to be some way for Verifiers to
+// check the correctness of the shares they receive. This is the scheme goal.
+// 1) The dealer send a Deal to every verifiers using `Deals()`
+// 2) Each verifier process the Deal with `ProcessDeal`. They can either return:
+//    - an Approval, which means the deal seems correct
+//    - an Complaint, which means the deal is not correct and the dealer might
+//    be malicious.
+// 3) The dealer can respond to each Complaint by a Justification, basically
+// revealing the share he sent out to the original "complainer"
+// 4) All verifiers accept the deal iif there has been at least t Approval. They
+// refuse the deal if there has been more than t-1 complaints OR if a
+// Justification is wrong.
+// At the end of the scheme, all verifiers can re-unite their Share to
+// reconstruct the original secret. Only t out of n verifiers are needed.
+package vss
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/share"
+	"github.com/dedis/crypto/sign"
+)
+
+// Dealer is responsible for creating the shares, distribute them and reply to
+// any Complaints.
+type Dealer struct {
+	suite abstract.Suite
+
+	reader cipher.Stream
+	long   abstract.Scalar
+
+	secret abstract.Scalar
+
+	verifiers []abstract.Point
+	// threshold of shares that is needed to reconstruct the secret
+	t int
+	// sessionID is a unique identifier for the whole session of the scheme
+	sessionID []byte
+
+	// list of deals this Dealer has generated
+	deals []*Deal
+	*aggregator
+}
+
+// Deal is sent by the Dealer to each participants.
```

Comment:
Personally, I would prefer to have the following layout (as it gives a better overview):
```
type foo struct {
   name1 type1 //comment1
   name2 type2 //comment2
} 
```
However, that's just a matter of personal preference.